web/courses.go

```diff
@@ -125,38 +125,99 @@ func (s *AutograderService) getSubmissions(request *pb.SubmissionRequest) (*pb.S
-
-	// make a local map to store database values to avoid querying the database multiple times
-	// format: [studentID][assignmentID]{latest submission}
-	labCache := make(map[uint64]map[uint64]pb.Submission)
+	switch request.Type {
+	case pb.SubmissionsForCourseRequest_GROUP:
 
-	// populate cache map with student labs, filtering the latest submissions for every assignment
-	for _, lab := range allLabs {
-		labID := lab.GetUserID()
-		if request.GroupLabs {
-			labID = lab.GetGroupID()
+		enrols, err := s.makeGroupResults(course, assignments)
+		if err != nil {
+			return nil, err
 		}
-		_, ok := labCache[labID]
-		if !ok {
-			labCache[labID] = make(map[uint64]pb.Submission)
+		enrolLinks = append(enrolLinks, enrols...)
+	case pb.SubmissionsForCourseRequest_INDIVIDUAL:
+		enrolLinks = append(enrolLinks, makeResults(course, assignments)...)
+	default:
+		grpLinks, err := s.makeGroupResults(course, assignments)
+		if err != nil {
+			return nil, err
 		}
-		labCache[labID][lab.GetAssignmentID()] = lab
+		enrolLinks = append(makeResults(course, assignments), grpLinks...)
 	}
+	return &pb.CourseSubmissions{Course: course, Links: enrolLinks}, nil
+}
 
-	// fetch course record with all assignments and active enrollments
-	course, err := s.db.GetCourse(request.GetCourseID(), true)
-	if err != nil {
-		return nil, err
+// makeResults generates enrollment to assignment to submissions links
+// for all course students and all individual assignments
+func makeResults(course *pb.Course, assignments []*pb.Assignment) []*pb.EnrollmentLink {
+	enrolLinks := make([]*pb.EnrollmentLink, 0)
+
+	for _, enrol := range course.Enrollments {
+		newLink := &pb.EnrollmentLink{Enrollment: enrol}
+		allSubmissions := make([]*pb.SubmissionLink, 0)
+		for _, a := range assignments {
+			subLink := &pb.SubmissionLink{
+				Assignment: a,
+			}
+			for _, sb := range a.Submissions {
+				if sb.UserID > 0 && sb.UserID == enrol.UserID {
+					subLink.Submission = sb
+				}
+			}
+			allSubmissions = append(allSubmissions, subLink)
+		}
+
+		sortSubmissionsByAssignmentID(allSubmissions)
+		newLink.Submissions = allSubmissions
+		enrolLinks = append(enrolLinks, newLink)
 	}
 
-	return makeLabResults(course, labCache, request.GetGroupLabs()), nil
+	return enrolLinks
+}
+
+// makeGroupResults generates enrollment to assignment to submissions links
+// for all course groups and all group assignments
+func (s *AutograderService) makeGroupResults(course *pb.Course, assignments []*pb.Assignment) ([]*pb.EnrollmentLink, error) {
+	enrolLinks := make([]*pb.EnrollmentLink, 0)
+	for _, grp := range course.Groups {
+
+		newLink := &pb.EnrollmentLink{}
+		for _, enrol := range course.Enrollments {
+			if enrol.GroupID > 0 && enrol.GroupID == grp.ID {
+				newLink.Enrollment = enrol
+			}
+		}
+		if newLink.Enrollment == nil {
+			s.logger.Debugf("Got empty enrollment for group %+v", grp)
```

Comment:
Okay, this is fine, but this method no longer needs to return an error, making it easier to use the result in `getAllLabs`.